File: backend/src/models/message.model.js
const mongoose = require('mongoose');

const messageSchema = new mongoose.Schema({
  sender: {
    type: mongoose.Schema.ObjectId,
    ref: 'User',
    required: [true, 'A message must have a sender']
  },
  receiver: {
    type: mongoose.Schema.ObjectId,
    ref: 'User',
    required: [true, 'A message must have a receiver']
  },
  content: {
    type: String,
    required: [true, 'Message content cannot be empty'],
    trim: true,
    maxlength: [1000, 'Message content cannot exceed 1000 characters']
  },
  relatedBook: {
    type: mongoose.Schema.ObjectId,
    ref: 'Book'
  },
  isRead: {
    type: Boolean,
    default: false
  },
  createdAt: {
    type: Date,
    default: Date.now
  }
});

// Create compound index for conversation querying
messageSchema.index({ sender: 1, receiver: 1 });

module.exports = mongoose.model('Message', messageSchema); 

File: lib/models/message_model.dart
class Message {
  final String id;
  final Map<String, dynamic> sender;
  final Map<String, dynamic> receiver;
  final String content;
  final Map<String, dynamic>? relatedBook;
  final bool isRead;
  final DateTime createdAt;

  Message({
    required this.id,
    required this.sender,
    required this.receiver,
    required this.content,
    this.relatedBook,
    required this.isRead,
    required this.createdAt,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['_id'],
      sender: json['sender'],
      receiver: json['receiver'],
      content: json['content'],
      relatedBook: json['relatedBook'],
      isRead: json['isRead'] ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'sender': sender,
      'receiver': receiver,
      'content': content,
      'relatedBook': relatedBook,
      'isRead': isRead,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}
